### PR TITLE
feat: support merge by row_id, row_addr

### PR DESCRIPTION
**Pull request by chenkovsky**

### CODE REVIEW

File: rust/lance/src/dataset.rs [hunk 1]
```diff
@@ -3117,6 +3118,111 @@ mod tests {
...
+        let new_schema = Arc::new(ArrowSchema::new(vec![
+            ArrowField::new("rowid", DataType::UInt64, false),
+            ArrowField::new("new_value", DataType::UInt64, false),
+        ]));
+        let new_batch =
+            RecordBatch::try_new(new_schema.clone(), vec![row_ids.clone(), row_ids.clone()])
+                .unwrap();
+        let new_data = RecordBatchIterator::new(vec![Ok(new_batch)], new_schema.clone());
+        dataset.merge(new_data, ROW_ID, "rowid").await.unwrap();
+        dataset.validate().await.unwrap();
```
Comment (wjones127) [inline rust/lance/src/dataset.rs:3168]: I'd like us to assert a few more things in this test:

1. `dataset` has the expected final schema `key, value, new_value`.
2. The values are what we expect. For this, you should avoid using the same values in each column. Otherwise, the test could pass even if there is a bug that uses the wrong column's values. Right now, you use `row_ids.clone()` for both `rowid` and `new_value`.
3. This works even if you shuffle the data. I would recommend using [take_record_batch()](https://docs.rs/arrow/latest/arrow/compute/fn.take_record_batch.html) to reorder the `new_batch` so the row ids are out-of-order.

File: rust/lance/src/dataset.rs [hunk 2]
```diff
@@ -3117,6 +3118,111 @@ mod tests {
...
+    #[tokio::test]
+    async fn test_merge_on_row_id(
+        #[values(LanceFileVersion::Legacy, LanceFileVersion::Stable)]
+        data_storage_version: LanceFileVersion,
+        #[values(false, true)] use_stable_row_id: bool,
+    ) {
+        // Tests a merge that spans multiple batches within files
+
+        // This test also tests "null filling" when merging (e.g. when keys do not match
+        // we need to insert nulls)
```
Comment (wjones127) [inline rust/lance/src/dataset.rs:3131]: Where is the null filling? It seems like you are providing every row id, unless I am missing something.

File: rust/lance/src/dataset.rs [hunk 3]
```diff
@@ -3117,6 +3118,111 @@ mod tests {
...
+        dataset.merge(new_data, ROW_ID, "rowid").await.unwrap();
+        dataset.validate().await.unwrap();
+    }
+
+    #[rstest]
+    #[tokio::test]
+    async fn test_merge_on_row_addr(
+        #[values(LanceFileVersion::Legacy, LanceFileVersion::Stable)]
+        data_storage_version: LanceFileVersion,
+        #[values(false, true)] use_stable_row_id: bool,
```
Comment (wjones127) [inline rust/lance/src/dataset.rs:3176]: Same comments from the row id test apply here.

**Review by wjones127 — CHANGES_REQUESTED** [review]
Thanks for working on this @chenkovsky. I would like to see a few improvements to the unit tests, and then this is ready to go.

File: rust/lance/src/dataset.rs (hunk 2 above)
Comment (chenkovsky) [reply rust/lance/src/dataset.rs:3131]: > Where is the null filling? It seems like you are providing every row id, unless I am missing something.

sorry, I copy and modify another test

File: rust/lance/src/dataset.rs (hunk 3 above)
Comment (chenkovsky) [reply rust/lance/src/dataset.rs]: updated

File: rust/lance/src/dataset.rs [hunk 4]
```diff
@@ -3117,6 +3119,182 @@ mod tests {
...
+        assert!(dataset.schema().field("value").is_some());
+        assert!(dataset.schema().field("new_value").is_some());
+        let result = dataset
+            .scan()
+            .try_into_stream()
+            .await
+            .unwrap()
+            .try_collect::<Vec<_>>()
+            .await
+            .unwrap();
```
Comment (wjones127) [inline rust/lance/src/dataset.rs:3278]: You can use `.try_into_batch()` to immediately collect the results in one batch. Will need to remove the for loop below too.
```suggestion
        let result = dataset
            .scan()
            .try_into_batch()
            .await
            .unwrap();
```

File: rust/lance/src/dataset.rs [hunk 5]
```diff
@@ -3117,6 +3119,182 @@ mod tests {
...
+        for batch in result {
+            let key = batch
+                .column_by_name("key")
+                .unwrap()
+                .as_any()
+                .downcast_ref::<arrow_array::Int32Array>()
+                .unwrap();
```
Comment (wjones127) [inline rust/lance/src/dataset.rs:3285]: For tests, you can use `RecordBatch["<column>"]` to access a column by name. Also, `.as_primitive()` is typically shorter. Both of these are mostly suitable for tests because they can panic.

```suggestion
            let key = batch["key"].as_primitive::<Int32Type>();
```

**Review by wjones127 — CHANGES_REQUESTED** [review]
Looks better. Have a few suggestions to clean up the tests. Then I think this is ready to be merged.

**Review by wjones127 — APPROVED** [review]
